984 Print LB specification at startup (from file)

cz4rs opened the pull request:
fixes #984

cz4rs commented on src/vt/vrt/collection/balance/read_lb.cc:
```diff
@@ -302,4 +301,15 @@ ReadLBSpec::parseParams(std::vector<std::string> params) {
   return SpecEntry{0, "", param_map};
 }
 
+/*static*/ SpecIndex ReadLBSpec::numEntries() {
```
`numEntries` was clearly not working (always returning 0), I'm trying to fix it here

questions:
- do we consider an empty file a valid configuration? we need to return 0 in that case (and add a unit test for that)
- do we want actual entries (in which case we should probably return `spec_mod_.size() + spec_exact_.size()`) or something more like "number of phases covered" (which is actually implemented here)

lifflander replied:
- I think an empty file is probably a user mistake and we should at the very least print a warning. In the runtime banner when we print the LB file, if it's empty maybe we can print a warning in red.
- I think `numEntries` should just be a actual count of entries in the file so `spec_mod_.size() + spec_exact_.size()`

cz4rs replied:
- `numEntries` fixed
- we already abort when the file is empty (while trying to read and parse the file):
```
vt: [0] ------------------------------------------- Runtime Error: System Aborting! --------------------------------------------
vt: [0] ------------------------------------------------ Fatal Error on Node 0 -------------------------------------------------
vt: [0] ------------------------------------------------------------------------------------------------------------------------
vt: [0] 
vt: [0]              Reason: Valid LB not found: "name="

vt: [0]                Type: vtAbort() Invoked
```
it's not necessarily clear for the user what's wrong, so I can add the warning as well

cz4rs commented on src/vt/runtime/runtime_banner.cc:
```diff
@@ -62,6 +63,26 @@
 
 #include <mpi.h>
 
+void printLBConfig(std::string const& filename) {
```
a prototype for further discussion, currently prints:
```
vt:     Option: flag --vt_lb on: Load balancing enabled
vt:     Option: flag --vt_lb_file on: Reading LB config from file
vt:     Option: flag --vt_lb_file_name on: Reading file "./lb_file"
Print LB configuration file here
i: 0    idx: 0  name: HierarchicalLB
i: 1
i: 2
i: 3
i: 4
i: 5    idx: 5  name: GreedyLB
i: 6
i: 7
i: 8
i: 9
i: 10   idx: 10 name: GossipLB
i: 11
i: 12
i: 13
i: 14
i: 15   idx: 5  name: GreedyLB
i: 16
i: 17
i: 18
i: 19
i: 20   idx: 20 name: GreedyLB
vt:     Option: flag --vt_print_no_progress on: Printing warnings when progress is stalls
```
- apart from making it "pretty", what information should be included in each line? (parameters?)
- should we skip the lines where `entry(i)` returns `nullptr`?

lifflander replied:
- For printing the configuration, we need to specify which lines override others. More exact ones override mods and the order determines which is applied if they conflict.
- I think we should print in in English.
```
Run `HierarchicalLB` on phase 20 with arguments `a=x b=y`
Run `GossipLB` every 10 phases excluding phase 20
```

Do you know why `entry(i)` would be a `nullptr`? Is that bug?

cz4rs replied:
`entry(i)` looks for specification at provided index (phase?) and returns `nullptr` if there is none

```cpp
  // Else, return nullptr---no applicable entry found
  return nullptr;
```
this seems to make sense, because we don't require specification file to cover all indexes

pnstickne replied:
If it's just printing the file, `cat filename` is more appropriate. If it's printing the configuration, perhaps:

    --vt_lb_print_config

cz4rs replied:
this specification:
```
%10 GossipLB c=1 k=5 f=2 i=10
0 HierarchicalLB min=0.9 max=1.1 auto=false
% 5 GreedyLB min=1.0
20 GreedyLB c=0 k=2 f=3 i=3
```
currently produces:
```
vt: Runtime Configuration:
vt:     Option: flag --vt_sched_num_progress on: Running MPI progress 2 times each invocation
vt:     Option: flag --vt_sched_progress_han on: Running MPI progress function at least every 0 handler(s) executed
vt:     Option: flag --vt_lb on: Load balancing enabled
vt:     Option: flag --vt_lb_file_name on: Reading LB specification from file "./lb_file"
vt:     Option: flag --vt_lb_show_spec on: Showing LB specification
vt:     Exact specification lines:
vt:     Run `HierarchicalLB` on phase 0 with arguments `auto=false max=1.1 min=0.9`
vt:     Run `GreedyLB` on phase 20 with arguments `i=3 f=3 k=2 c=0`
vt:     Mod (%) specification lines:
vt:     Run `GreedyLB` every 5 phases with arguments `min=1.0`
vt:     Run `GossipLB` every 10 phases with arguments `i=10 f=2 k=5 c=1`
vt:     Option: flag --vt_print_no_progress on: Printing warnings when progress is stalls
```
![image](https://user-images.githubusercontent.com/4639165/90251545-34006080-de3e-11ea-9b2d-ca2caac85ea5.png)

cz4rs replied:
after small refactoring, this seems to be ready to go:
```
vt: Runtime Configuration:
vt:     Option: flag --vt_sched_num_progress on: Running MPI progress 2 times each invocation
vt:     Option: flag --vt_sched_progress_han on: Running MPI progress function at least every 0 handler(s) executed
vt:     Option: flag --vt_lb on: Load balancing enabled
vt:     Option: flag --vt_lb_file_name on: Reading LB specification from file "./lb_file"
vt:     Option: flag --vt_lb_show_spec on: Showing LB specification
vt:     Exact specification lines:
vt:     Run `HierarchicalLB` on phase 0 with arguments `auto=false max=1.1 min=0.9`
vt:     Run `GreedyLB` on phase 20 with arguments `c=0 f=3 i=3 k=2`
vt:     Mod (%) specification lines:
vt:     Run `GreedyLB` every 5 phases with arguments `min=1.0` excluding phases 10, 20
vt:     Run `GossipLB` every 10 phases with arguments `c=1 f=2 i=10 k=5` excluding phases 20
vt:     Option: flag --vt_print_no_progress on: Printing warnings when progress is stalls
```
![image](https://user-images.githubusercontent.com/4639165/90298287-fb8a7200-de91-11ea-9bf4-6e6b8b3d04af.png)

lifflander replied:
This summary looks great to me!

cz4rs commented:
@lifflander can you please have a look at the sample output and the comments above?

also, minor improvements possible:
- `--vt_lb_file_name` usage string is "LB configuration file to read", html documentation mentions "LB specification file" - should we make this consistent?
- it doesn't seem to be possible to use `--vt_lb_file_name` without enabling `--vt_lb_file`, maybe we can remove the latter as redundant? (and just check if `--vt_lb_file_name`is valid as we do already in `ReadLBSpec::hasSpec()`

lifflander commented:
> @lifflander can you please have a look at the sample output and the comments above?
> 
> also, minor improvements possible:
> 
> * `--vt_lb_file_name` usage string is "LB configuration file to read", html documentation mentions "LB specification file" - should we make this consistent?

Yes, let's make it consistent with the documentation.

> * it doesn't seem to be possible to use `--vt_lb_file_name` without enabling `--vt_lb_file`, maybe we can remove the latter as redundant? (and just check if `--vt_lb_file_name`is valid as we do already in `ReadLBSpec::hasSpec()`

Yes, I've been thinking about removing the redundant flag, let's remove it.

cz4rs commented:
this should look better now, things left to do:
- specify which lines override others
- color output
- at least some doxygen comments
- cleanup and unit tests where aprropriate

pnstickne commented on src/vt/configs/arguments/args.cc:
```diff
@@ -342,9 +342,10 @@ void ArgConfig::addLbArgs(CLI::App& app) {
-  auto u  = app.add_option("--vt_lb_file_name",  config_.vt_lb_file_name,   lb_file_name, lbf);
+  auto u  = app.add_option("--vt_lb_file_name",  config_.vt_lb_file_name,   lb_file_name, lbf)
+    ->check(CLI::ExistingFile);
+  auto u1 = app.add_flag("--vt_lb_print_file",    config_.vt_lb_print_file,  lb_print_file);
```
If it's just display a 'file', `cat filename` is more appropriate. If it's printing the configuration, perhaps:

    --vt_lb_show_config

Notes:
- Config, not file
- Show is the action, print is the implementation

cz4rs replied:
agreed, going with `--vt_lb_show_spec` for consistency in the end (as decided in "configuration" vs "specification" above)

lifflander replied:
`--vt_lb_show_spec` sounds good to me! Let's go with that.

pnstickne commented on src/vt/runtime/runtime_banner.cc:
```diff
@@ -62,6 +63,32 @@
+    fmt::print("Run `{}` on phase {} with arguments `{}`\n", exact_entry.second.getName(), exact_entry.second.getIdx(), formatted_params);
+  }
+  for (auto const& mod_entry : Spec::getModEntries()) {
+        std::string formatted_params;
```
Indenting.

cz4rs replied:
fixed

pnstickne commented on src/vt/runtime/runtime_banner.cc:
```diff
@@ -62,6 +63,32 @@
+      formatted_params += fmt::format("{}={} ", param.first, param.second);
+    }
+
+    fmt::print("Run `{}` on phase {} with arguments `{}`\n", exact_entry.second.getName(), exact_entry.second.getIdx(), formatted_params);
```
Split line - consolidate effort to be <= 80 chars.

In this case the values drop-down nicely.

cz4rs replied:
done

pnstickne commented on src/vt/runtime/runtime_banner.cc:
```diff
@@ -62,6 +63,32 @@
+  for (auto const& exact_entry : Spec::getExactEntries()) {
+    std::string formatted_params;
+    for (auto const& param : exact_entry.second.getParams()) {
+      formatted_params += fmt::format("{}={} ", param.first, param.second);
```
This might be easier to emit incrementally instead of building up the string before.

Using a stringstream might be helpful.

cz4rs replied:
done

pnstickne commented on src/vt/vrt/collection/balance/read_lb.cc:
```diff
@@ -57,35 +57,28 @@
   }
+
+  auto const file_name = theConfig()->vt_lb_file_name;
+  if (file_name == "") {
```
Should match the if-above.
eg.

Both:

      if (x)
         return y;

Or

     if (x) {
         return t;
     }

Strive for consistency, working out.

cz4rs replied:
this file already uses "always braces" style consistently (I think it's ok as it is?)

lifflander replied:
I'm not sure if I understand @pnstickne comment. Can you elaborate?

pnstickne commented on src/vt/configs/arguments/app_config.h:
```diff
@@ -130,7 +130,7 @@ struct AppConfig {
   bool vt_trace_irecv_polling    = false;
 
   bool vt_lb                  = false;
-  bool vt_lb_file             = false;
```
Nice simpler model of flag detection.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/lb/test_lb_reader.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6007507&bid=19693488)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/lb/test_lb_reader.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6007507&bid=19693488)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/lb/test_lb_reader.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6007507&bid=19693488)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/unit/lb/test_lb_reader.cc  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6007507&bid=19693488)

cz4rs commented:
- reorganized the code and moved it away from `runtime_banner.cc`
- covered (most of?) new code with unit tests
- switched from `std::unordered_map` to `std::map` in `read_lb.h` (to make my life easier and have things sorted, hopefully this is not performance critical)

lifflander commented:
> * reorganized the code and moved it away from `runtime_banner.cc`
> * covered (most of?) new code with unit tests
> * switched from `std::unordered_map` to `std::map` in `read_lb.h` (to make my life easier and have things sorted, hopefully this is not performance critical)

Great! Using `std::map` shouldn't be a problem at all.
